File: frontend/src/types/holding.ts
/**
 * TypeScript types that mirror the backend Pydantic schemas.
 *
 * Keep these in sync with:
 *   backend/app/schemas/holding.py
 */

/** Payload sent to POST /api/v1/holdings */
export type HoldingCreate = {
  ticker: string;
  quantity_added: string;
  purchase_price: string;
  purchase_date: string;
};

/** Partial payload for PATCH /api/v1/holdings/:id */
export type HoldingUpdate = {
  ticker?: string;
  quantity_added?: string;
  purchase_price?: string;
  purchase_date?: string;
};

/** Single holding returned by the API (includes server-assigned id). */
export type Holding = {
  id: number;
  ticker: string;
  quantity_added: string;
  purchase_price: string;
  purchase_date: string | null;
};

/** Holding with live price and performance metrics from GET /holdings/performance */
export type HoldingPerformance = Holding & {
  current_price: string;
  cost_basis: string;
  market_value: string;
  gain_loss: string;
  gain_loss_percentage: string | null;
};

/** Reserved for a future portfolio summary endpoint. */
export type PortfolioSummary = {
  total_cost_basis: string;
  total_market_value: string;
  total_gain_loss: string;
  portfolio_return_percentage: string;
};

File: backend/app/schemas/holding.py

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, ValidationError, Field

class HoldingCreate(BaseModel):
    ticker: str = Field(min_length=1, max_length=20)
    quantity_added: Decimal = Field(gt=0)
    purchase_price: Decimal = Field(ge=0)
    purchase_date: date

class HoldingUpdate(BaseModel):
    ticker: str | None = Field(default=None, min_length=1, max_length=20)
    quantity_added: Decimal | None = Field(default=None, gt=0)
    purchase_price: Decimal | None = Field(default=None, ge=0)
    purchase_date: date | None = None

class HoldingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    ticker: str
    quantity_added: Decimal
    purchase_price: Decimal
    purchase_date: date | None = None


class HoldingPerformanceResponse(HoldingResponse):
    current_price: Decimal
    cost_basis: Decimal
    market_value: Decimal
    gain_loss: Decimal
    gain_loss_percentage: Decimal | None

class PortfolioSummaryResponse(BaseModel):
    total_cost_basis: Decimal
    total_market_value: Decimal
    total_gain_loss: Decimal
    portfolio_return_percentage: Decimal